Response header configuration

mrsaldana opened the pull request:
fixes #18108

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
\ No newline at end of file
+cookies.samesite.knownDuplicateName.useraction=Ensure that cookie names and patterns are unique across the 'lax', 'none', and 'strict' configuration lists.
+
+headers.emptyName=CWWKT0038W: An empty header name was found while parsing the '{0}' configuration. This value will be ignored.
```
Update

>An empty header name was found while parsing the '{0}' configuration. This value will be ignored.

to

>An empty header name was found when the '{0}' configuration was parsed. This value is ignored.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
+headers.emptyName.explanation=The specified header's name is an empty string. A header name must not be empty. 
+headers.emptyName.useraction=Ensure that no header name is empty.  
+
+headers.duplicateHeaderName=CWWKT0039W: A duplicate header name was found in the header [{0}] using the 'set' configuration. All configurations for the [{0}] header will be ignored. Any header defined by the 'set' configuration must contain unique header names.
```
Update

>A duplicate header name was found in the header [{0}] using the 'set' configuration. All configurations for the [{0}] header will be ignored. Any header defined by the 'set' configuration must contain unique header names.

to

>A duplicate header name was found in the [{0}] header using the 'set' configuration. All configurations for the [{0}] header are ignored. Any header that is defined by the 'set' configuration must contain unique header names.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
+headers.emptyName.useraction=Ensure that no header name is empty.  
+
+headers.duplicateHeaderName=CWWKT0039W: A duplicate header name was found in the header [{0}] using the 'set' configuration. All configurations for the [{0}] header will be ignored. Any header defined by the 'set' configuration must contain unique header names.
+headers.duplicateHeaderName.explanation=The specified header name was already found by the 'set' and/or 'add' configurations.
```
Update 

>The specified header name was already found by the 'set' and/or 'add' configurations.

to

>The specified header name was already found by the 'set' or 'add' configurations.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
+
+headers.duplicateHeaderName=CWWKT0039W: A duplicate header name was found in the header [{0}] using the 'set' configuration. All configurations for the [{0}] header will be ignored. Any header defined by the 'set' configuration must contain unique header names.
+headers.duplicateHeaderName.explanation=The specified header name was already found by the 'set' and/or 'add' configurations.
+headers.duplicateHeaderName.useraction=Ensure that header names defined on 'set' configuration are unique across the 'add' and 'set' configurations.
```
Update

>Ensure that header names defined on 'set' configuration are unique across the 'add' and 'set' configurations.

to

>Ensure that header names that are defined in the 'set' configuration are unique across the 'add' and 'set' configurations.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
+headers.duplicateHeaderName.explanation=The specified header name was already found by the 'set' and/or 'add' configurations.
+headers.duplicateHeaderName.useraction=Ensure that header names defined on 'set' configuration are unique across the 'add' and 'set' configurations.
+
+headers.knownDuplicateHeader=CWWKT0040W: The header [{0}], which is marked as duplicate header name, was found in the 'set' configuration. The [{0}] header is ignored. Any header defined by the 'set' configuration must contain unique header names.
```
Update

>The header [{0}], which is marked as duplicate header name, was found in the 'set' configuration. The [{0}] header is ignored. Any header defined by the 'set' configuration must contain unique header names.

to

>The [{0}] header, which is marked as a duplicate header name, was found in the 'set' configuration. The [{0}] header is ignored. Any header that is defined by the 'set' configuration must contain unique header names.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
+headers.duplicateHeaderName.useraction=Ensure that header names defined on 'set' configuration are unique across the 'add' and 'set' configurations.
+
+headers.knownDuplicateHeader=CWWKT0040W: The header [{0}], which is marked as duplicate header name, was found in the 'set' configuration. The [{0}] header is ignored. Any header defined by the 'set' configuration must contain unique header names.
+headers.knownDuplicateHeader.explanation=The specified header name was already found by the 'set' and/or 'add' configurations.
```
Update

>The specified header name was already found by the 'set' and/or 'add' configurations.

to

>The specified header name was already found by the 'set' or 'add' configurations.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -176,4 +176,16 @@ cookies.samesite.unsupportedWildcard.useraction=Ensure that only a single wildca
+
+headers.knownDuplicateHeader=CWWKT0040W: The header [{0}], which is marked as duplicate header name, was found in the 'set' configuration. The [{0}] header is ignored. Any header defined by the 'set' configuration must contain unique header names.
+headers.knownDuplicateHeader.explanation=The specified header name was already found by the 'set' and/or 'add' configurations.
+headers.knownDuplicateHeader.useraction=Ensure that header names defined on 'set' configuration are unique across the 'add' and 'set' configurations.               
```
Update

>Ensure that header names defined on 'set' configuration are unique across the 'add' and 'set' configurations.

to

>Ensure that header names that are defined in the 'set' configuration are unique across the 'add' and 'set' configurations.

andrea-pichler commented:
Completed the requested L2 message review

mrsaldana commented:
@Charlotte-Holt I've added the suggested changes. I also added some documentation on the metatype regarding the two attributes being added and the new element that holds them. If you have additional suggestions on these, I'm more than happy to apply them. Thanks in advanced!

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -232,4 +239,10 @@ http.samesite.none=List of samesite none cookies
\ No newline at end of file
+http.samesite.strict.desc=List of cookie names or patterns for which the SameSite attribute is set to a value of Strict, if not already defined. A single wildcard (*) character is supported as a stand-alone value, or following cookie name prefixes. Any cookie name or pattern that is defined by this list must be unique and not present in the 'lax' nor 'none' configurations.
+
+http.headers.add=Headers to be added on each responses
```
Update

>Headers to be added on each responses

to

>Headers to be added on each response

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -232,4 +239,10 @@ http.samesite.none=List of samesite none cookies
+http.samesite.strict.desc=List of cookie names or patterns for which the SameSite attribute is set to a value of Strict, if not already defined. A single wildcard (*) character is supported as a stand-alone value, or following cookie name prefixes. Any cookie name or pattern that is defined by this list must be unique and not present in the 'lax' nor 'none' configurations.
+
+http.headers.add=Headers to be added on each responses
+http.headers.add.desc=Specifies the header names and values that will be added to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name defined using this attribute must not be empty. 
```
Update

>Specifies the header names and values that will be added to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name defined using this attribute must not be empty. 

to

>Specifies the header names and values that are added to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name that is defined by using this attribute must not be empty.

Charlotte-Holt commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -232,4 +239,10 @@ http.samesite.none=List of samesite none cookies
+http.headers.add.desc=Specifies the header names and values that will be added to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name defined using this attribute must not be empty. 
+
+http.headers.set=Headers to be set on each response
+http.headers.set.desc=Specifies the header names and values that will be set to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name defined using this attribute must not be empty, unique, and not present in the 'add' header configuration.
```
Update

>Specifies the header names and values that will be set to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name defined using this attribute must not be empty, unique, and not present in the 'add' header configuration.

to

>Specifies the header names and values that are set to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name that is defined by using this attribute must not be empty or unique, and it must not present in the 'add' header configuration.

helyarp commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -101,6 +105,9 @@ http.compression.desc=An element that is configured within the httpEndpoint elem
 http.samesite.desc=An element that is configured within the httpEndpoint element so that the associated HTTP channel can consider SameSite configurations.
 
+http.headers=Header Options
+http.headers.desc=An element that is configured within the httpEndpoint element so that the associated HTTP channel can consider response header configurations.
```
Suggestion: Replace "consider" with another verb, maybe "evaluate," "assess," "examine," or "access."  Normally, "consider" is something that humans/animals do.

billw1024 reviewed (approved):
nice job

helyarp commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -232,4 +239,16 @@ http.samesite.none=List of samesite none cookies
+http.headers.add.desc=Specifies the header names and values that are added to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name that is defined by using this attribute must not be empty or present in the 'remove', 'set', or 'setIfMissing' header configurations. 
+
+http.headers.set=Headers to be set on each response
+http.headers.set.desc=Specifies the header names and values that are set to each HTTP response. Header values are optionally included by using a colon (:) delimiter. Any header name that is defined by using this attribute must not be empty, defined more than once, or present in the 'remove', 'add', or 'setIfEmpty'  header configurations. If the header is already present on the response, existing  ...
```
Change "values will be" to "values are"

helyarp commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -166,14 +166,26 @@ decompression.tolerance.reached=CWWKT0034E: The HTTP channel attempted to decomp
 cookies.samesite.knownDuplicateName.useraction=Ensure that cookie names and patterns are unique across the 'lax', 'none', and 'strict' configuration lists.
+
+headers.emptyName=CWWKT0038W: An empty header name was found when the '{0}' configuration was parsed. This value is ignored.
+headers.emptyName.explanation=The specified header's name is an empty string. A header name must not be empty. 
```
Remove the 's because it is not needed and it can adversely affect translation. Use "specified header name" instead.

helyarp commented on dev/com.ibm.ws.transport.http/resources/com/ibm/ws/http/channel/internal/resources/httpchannelmessages.nlsprops:
```diff
@@ -166,14 +166,26 @@ decompression.tolerance.reached=CWWKT0034E: The HTTP channel attempted to decomp
+
+headers.duplicateHeaderName=CWWKT0039W: A duplicate header name was found in the [{0}] header using the '{1}' configuration. All configurations for the [{0}] header are ignored. Any header that is defined by the 'remove', 'add', 'set', or 'setIfMissing' configurations must be unique across all configurations.
+headers.duplicateHeaderName.explanation=The specified header name was previously configured by another configuration of the Headers feature.
+headers.duplicateHeaderName.useraction=Ensure that headers that are defined by the 'remove', 'add', 'set', or 'setIfMissing' configurations are no present in more than one configuration.
```
Change "are no present" to "are not present"

helyarp reviewed (commented):
A few suggested changes

helyarp reviewed (commented):
Looks good!

mrsaldana commented:
#build

pnicolucci commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -88,9 +88,9 @@
-            type="String" required="false" default="defaultHttpOptions" />
+        <AD id="httpOptionsRef" name="%endpoint.httpOptions"
+	description="%endpoint.httpOptions.desc" ibm:reference="com.ibm.ws.http.options"
+	ibm:type="pid" type="String" required="false" default="defaultHttpOptions" />
```
Can you clean up the formatting here?

mrsaldana replied:
Formatting on this entire file looks fine on a text editor, but shows up messed up on the git preview. My guess is that there are 'tab' and whitespace characters that look fine in the eclipse or regular text editors but get messed up on this git view. I propose opening up a separate PR to clean up the entire file, since the formatting seems off for multiple other segments as well.

pnicolucci replied:
https://github.com/OpenLiberty/open-liberty/issues/18107

pnicolucci commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -123,6 +127,10 @@
     
+    	<AD id="headers.target" name="internal" description="internal use only"
+			ibm:final="true" required="false" type = "String" ibm:beta="true"
+			default="(|(service.pid=${headersRef})(&amp;(service.factoryPid=com.ibm.ws.http.headers)(|(id=com.ibm.ws.http-${id}/*)(id=defaultHeaders))))" />
```
Formatting seems a bit odd here as well.

mrsaldana replied:
For this one, all the other elements have the same format, so leaving this one as is for now.

mrsaldana replied:
See the other metatype.xml comment for proposed solution.

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpConfigConstants.java:
```diff
@@ -437,6 +437,16 @@ private HttpConfigConstants() {
     public static final String WILDCARD_CHAR = "*";
 
+    public static final String PROPNAME_HEADERS_FEATURE = "headersInternal";
+
```
I'm not a huge fan of having the word `feature` in this naming convention. Can you work to come up with a better name here?

mrsaldana replied:
Changed to `PROPNAME_RESPONSE_HEADERS`

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpResponseMessageImpl.java:
```diff
@@ -711,6 +712,36 @@ public void headerComplianceCheck() throws MessageSentException {
 
+        if (getServiceContext().getHttpConfig().useHeadersConfiguration()) {
+            //Add all headers configured through the ADD configuration option
+            for (List<Map.Entry<String, String>> headers : getServiceContext().getHttpConfig().getConfiguredHeadersToAppend().values()) {
```
Any reason you use `append` here rather than `add`?

mrsaldana replied:
The idea was to match the `appendHeader()` API, but I have no objection for it to be inline with the configuration `add` format.

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/FATSuite.java:
```diff
@@ -87,7 +88,8 @@
                 WC5GetContextPath.class,
-                WCSCIHandlesTypesTest.class
+                WCSCIHandlesTypesTest.class,
+		 WCResponseHeadersTest.class
```
Please fix the formatting here.

mrsaldana replied:
Done

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java:
```diff
@@ -0,0 +1,1490 @@
+        String consoleLogFileName = WCResponseHeadersTest.class.getSimpleName() + ".log";
+        
+        if(!server.isStarted()){
+            server.updateServerConfiguration(savedConfig);
```
Add check to ensure savedConfig isn't null? What happens if it is null?

mrsaldana replied:
`savedConfig` is initialized by the `@BeforeClass `, where the `server` object is presumed to had already been initialized by the  runtime `@Server` annotation . I don't believe this should ever be null since it remains unchanged throughout the lifecycle of the test. If it were null, then each test would likely fail due to whatever configuration the server had on the previous test run. The sample documentation on the git wiki didn't suggest adding null checks on the `@BeforeClass` for the server, but it is a simple/quick check that. I'll add them, but I almost prefer the test fail and we find an exception that says that the setup failed, than it fail later on using a bad configuration and we end up having to debug to find that the issue wasn't in the test but on the configuration. 

It probably will be fairly obvious with how the tracing is setup, but some additional thoughts to consider.

pnicolucci replied:
If you think you're better off without the null check I don't have any objection given the details you've provided.

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java:
```diff
@@ -0,0 +1,1490 @@
+        if(!server.isStarted()){
+            server.updateServerConfiguration(savedConfig);
+            server.startServer(consoleLogFileName);
+            Log.info(ME, "setUpBeforeTest", "server started, log file is " + consoleLogFileName);
```
Perhaps use a `methodName` variable here to ensure the correct name. We should update this at the very least to actually match the method name we're in `setUpBeforeEachTest`.

mrsaldana replied:
Done, thanks for catching that one

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java:
```diff
@@ -0,0 +1,1490 @@
+            Log.info(ME, "setUpBeforeTest", "server started, log file is " + consoleLogFileName);
+        } else if(restoreSavedConfig){
+            try{
+                server.stopServer("CWWKT0042W", "CWWKT0043W", "CWWKT0044W");
```
Can we add comments here to detail what these warnings are? It makes it easier in the future to know why we're allowing these.

mrsaldana replied:
Done

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java:
```diff
@@ -0,0 +1,1490 @@
+            }
+            server.updateServerConfiguration(savedConfig);
+            server.startServer(consoleLogFileName, true);
+            Log.info(getClass(), "setUpBeforeTest", "server restarted, log file is " +consoleLogFileName);
```
In these start server calls do we need to ensure we're waiting for the config updates to complete? Something like:  `server.waitForConfigUpdateInLogUsingMark(Collections.singleton(appName), false/true, "CWWKT0016I:.*myAppName.*");`

mrsaldana replied:
I don't believe this would work since it is not updating the config on a running server, rather starting one up that is shutdown. The application is verified to be started through the `startServer` API, which does a `startServerAndValidate` with `validateApps` set to true by default. As such, I don't think we need to wait additional time after the `startServer` call returns.

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java:
```diff
@@ -0,0 +1,1490 @@
+            // W CWWKT0044W: The [{0}] header, which is marked as a duplicate header name, was found in the {1} configuration.
+            // The [{0}] header is ignored. Any header that is defined by the {1} configuration must contain unique header names.
+            if(server.isStarted())
+                server.stopServer("CWWKT0042W", "CWWKT0043W", "CWWKT0044W");
```
Usually we add a `server != null` check here as well.

mrsaldana replied:
These lines are already wrapped by a `if (server !=null)` block, around line 123 (give or take a few lines since it might be adjusted due to the other suggested changes.

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java:
```diff
@@ -0,0 +1,1490 @@
+        Header expectedHeader = null;
+
+        //server.saveServerConfiguration();
+
```
Can this be removed?

mrsaldana replied:
Done

mrsaldana commented:
#build

pnicolucci reviewed (approved):
Good work. Just ensure you get a good personal build with the latest changes.

mrsaldana commented:
#build

mrsaldana commented:
#build

mrsaldana commented:
#build

pnicolucci commented:
#run-libby-bot
